[SanitizerCoverage] Add an option to gate the invocation of the tracing callbacks

thetruestblue opened the pull request:
Implement -sanitizer-coverage-gated-trace-callbacks to gate the invocation of the tracing callbacks based on the value of a global variable, which is stored in a specific section.
When this option is enabled, the instrumentation will not call into the runtime-provided callbacks for tracing, thus only incurring in a trivial branch without going through a function call. It is up to the runtime to toggle the value of the global variable in order to enable tracing.

This option is only supported for trace-pc-guard. 

Note: will add additional support for trace-cmp in a follow up PR.

Patch by Filippo Bigarella

rdar://101626834

thetruestblue commented:
pinging for a review.

yln reviewed (approved):
LGTM, looks harmless enough.  Please ping again for review and wait a few more days before merging to give other reviewers another chance to chime in.

In the follow-up PR that expands support for this to trace-cmp, does it make sense to document this option here [SanitizerCoverage](https://clang.llvm.org/docs/SanitizerCoverage.html)?

thetruestblue committed:
[SanitizerCoverage] Add an option to gate the invocation of the tracing callbacks

Implement -sanitizer-coverage-gated-trace-callbacks to gate the invocation of the tracing callbacks based on the value of a global variable, which is stored in a specific section.
When this option is enabled, the instrumentation will not call into the runtime-provided callbacks for tracing, thus only incurring in a trivial branch without going through a function call. It is up to the runtime to toggle the value of the global variable in order to enable tracing.

This option is only supported for trace-pc-guard.

Note: will add additional support for trace-cmp in a follow up PR.

Patch by Filippo Bigarella

rdar://101626834
